tests/anoph/conftest.py

```diff
@@ -584,6 +584,318 @@ def simulate_aim_variants(path, contigs, snp_sites, n_sites_low, n_sites_high):
+        contig_grp.create_dataset(name="samples", data=samples, dtype=str)
+
+        # Create variants group for this contig.
+        variants_grp = contig_grp.require_group("variants")
+
+        # Simulate POS under variants.
+        # Note: choosing a random subset would cause dimension size conflicts.
+        variants_grp.create_dataset(name="POS", data=snp_pos)
+
+        # Simulate END under variants.
+        random_multipliers = np.random.randint(2, 12, size=len(snp_pos))
+        end = random_multipliers * snp_pos
+        variants_grp.create_dataset(name="END", data=end)
+
+    zarr.consolidate_metadata(zarr_path)
+
+
+def simulate_cnv_coverage_calls(zarr_path, metadata_path, contigs, snp_sites):
+    # zarr_path is the output path to the zarr store
+    # metadata_path is the input path for the sample metadata
+    # contigs is the list of contigs, e.g. Ag has ('2L', '2R', '3R', '3L', 'X')
+    # snp_sites are the data from snp_genotypes/all/sites, inc. {contig}/variants/POS
+    # n_sites will be derived from snp_sites {contig}/variants/POS
+
+    # {release}/cnv/{sample_set}/coverage_calls/{analysis}/zarr
+    #   - samples [1D array] [str for n_samples]
+    #   - {contig}
+    #      - calldata
+    #         - GT [2D array] [int] [0 or 1 for n_sites for n_samples]
+    #      - samples [1D array] [str for n_samples]
+    #      - variants
+    #         - CIEND [1D array] [int] [0 to 13200+ for n_sites]
+    #         - CIPOS [1D array] [int] [0 to 37200+ for n_sites]
+    #         - END [1D array] [int for n_sites]
+    #         - FILTER_PASS [1D array] [bool] [True or False for n_sites]
+    #         - FILTER_qMerge [1D array] [bool] [True or False for n_sites]
+    #         - ID [1D array] [unique str for n_sites]
+    #         - POS [1D array] [int for n_sites]
+
+    # Get a random probability for choosing allele 1, between 0 and 1.
+    p_allele = np.random.random()
+
+    # Get a random probability for passing a particular SNP site (position), between 0 and 1.
+    p_filter_pass = np.random.random()
+
+    # Get a random probability for applying qMerge filter to a particular SNP site (position), between 0 and 1.
+    p_filter_qMerge = np.random.random()
+
+    # Open a zarr at the specified path.
+    root = zarr.open(zarr_path, mode="w")
+
+    # Create samples array.
+    df_samples = pd.read_csv(metadata_path)
+    n_samples = len(df_samples)
+    samples = df_samples["sample_id"].values
+    root.create_dataset(name="samples", data=samples, dtype=str)
+
+    for contig in contigs:
+        # Create the contig group.
+        contig_grp = root.require_group(contig)
+
+        # Create the calldata group for this contig.
+        calldata = contig_grp.require_group("calldata")
+
+        # Get the SNP positions for this contig.
+        snp_pos = snp_sites[f"{contig}/variants/POS"][:]
+
+        # Get the number of sites (SNP positions) for this contig.
+        n_sites = snp_pos.shape[0]
```

Comment:
The CNV coverage calls are also completely independent of the SNP data, and so would be good to simulate them as such.

For these data, each variant is a CNV allele, which can start and end at an arbitrary position in the genome.

To simulate this, you could first randomly decide how many CNV variants to simulate, then simulate their start positions as random integers less than the contig size, then simulate their lengths as random integers in some range. You would also have to sort them so the CNVs appear in ascending order of start position.

POS should give the CNV start coordinate, and END should give the end coordinate.

Reply:
Many thanks again. I wonder what a good maximum number of CNV variants to simulate would be, i.e. to randomly simulate between 1 and N variants?

Reply:
Maybe somewhere between 1000 and 10000?

Reply:
Thanks. Have done.